// thirdweb/dist/esm/extensions/prebuilts/deploy-marketplace.js
import { toFunctionSelector, toFunctionSignature } from "viem";
import { resolveContractAbi } from "../../contract/actions/resolve-abi.js";
import { deployViaAutoFactory } from "../../contract/deployment/deploy-via-autofactory.js";
import { getOrDeployInfraContract, getOrDeployInfraForPublishedContract, } from "../../contract/deployment/utils/bootstrap.js";
import { upload } from "../../storage/upload.js";
import { getRoyaltyEngineV1ByChainId } from "../../utils/royalty-engine.js";
import { initialize as initMarketplace } from "./__generated__/Marketplace/write/initialize.js";
/**
 * Deploys a marketplace contract.
 * @param options - The options for deploying the marketplace contract.
 *
 * @extension MARKETPLACE
 *
 * @example
 * ```ts
 * import { deployMarketplaceContract } from "thirdweb/deploys";
 *
 * const address = await deployMarketplaceContract({
      client,
      chain,
      account,
      params: {
        name: "MarketplaceV3",
        description: "MarketplaceV3 deployed using thirdweb SDK",
        platformFeeRecipient: "0x21d514c90ee4E4e4Cd16Ce9185BF01F0F1eE4A04",
        platformFeeBps: 1000,
      },
    });
 * ```
 */
export async function deployMarketplaceContract(options) {
    const { chain, client, account, params, version } = options;
    const WETH = await getOrDeployInfraContract({
        chain,
        client,
        account,
        contractId: "WETH9",
    });
    const direct = await getOrDeployInfraForPublishedContract({
        chain,
        client,
        account,
        contractId: "DirectListingsLogic",
        constructorParams: { _nativeTokenWrapper: WETH.address },
    });
    const english = await getOrDeployInfraForPublishedContract({
        chain,
        client,
        account,
        contractId: "EnglishAuctionsLogic",
        constructorParams: { _nativeTokenWrapper: WETH.address },
    });
    const offers = await getOrDeployInfraForPublishedContract({
        chain,
        client,
        account,
        contractId: "OffersLogic",
    });
    const [directFunctions, englishFunctions, offersFunctions] = await Promise.all([
        resolveContractAbi(direct.implementationContract).then(generateExtensionFunctionsFromAbi),
        resolveContractAbi(english.implementationContract).then(generateExtensionFunctionsFromAbi),
        resolveContractAbi(offers.implementationContract).then(generateExtensionFunctionsFromAbi),
    ]);
    const { cloneFactoryContract, implementationContract } = await getOrDeployInfraForPublishedContract({
        chain,
        client,
        account,
        contractId: "MarketplaceV3",
        constructorParams: {
            _marketplaceV3Params: {
                extensions: [
                    {
                        metadata: {
                            name: "Direct Listings",
                            metadataURI: "",
                            implementation: direct.implementationContract.address,
                        },
                        functions: directFunctions,
                    },
                    {
                        metadata: {
                            name: "English Auctions",
                            metadataURI: "",
                            implementation: english.implementationContract.address,
                        },
                        functions: englishFunctions,
                    },
                    {
                        metadata: {
                            name: "Offers",
                            metadataURI: "",
                            implementation: offers.implementationContract.address,
                        },
                        functions: offersFunctions,
                    },
                ],
                royaltyEngineAddress: getRoyaltyEngineV1ByChainId(chain.id),
                nativeTokenWrapper: WETH.address,
            },
        },
        version,
    });
    const initializeTransaction = await getInitializeTransaction({
        client,
        implementationContract,
        params,
        accountAddress: account.address,
    });
    return deployViaAutoFactory({
        client,
        chain,
        account,
        cloneFactoryContract,
        initializeTransaction,
    });
}
async function getInitializeTransaction(options) {
    const { client, implementationContract, params, accountAddress } = options;
    const contractURI = options.params.contractURI ||
        (await upload({
            client,
            files: [
                {
                    name: params.name,
                    description: params.description,
                    image: params.image,
                    external_link: params.external_link,
                    social_urls: params.social_urls,
                },
            ],
        })) ||
        "";
    return initMarketplace({
        contract: implementationContract,
        contractURI,
        defaultAdmin: params.defaultAdmin || accountAddress,
        platformFeeBps: params.platformFeeBps || 0,
        platformFeeRecipient: params.platformFeeRecipient || accountAddress,
        trustedForwarders: params.trustedForwarders || [],
    });
}
// helperFns
function generateExtensionFunctionsFromAbi(abi) {
    const functions = abi.filter((item) => item.type === "function" && !item.name.startsWith("_"));
    return functions.map((fn) => ({
        functionSelector: toFunctionSelector(fn),
        functionSignature: toFunctionSignature(fn),
    }));
}
